streaming/src/main/scala/org/apache/spark/streaming/dstream/EmittedRecordsDStream.scala

```diff
@@ -0,0 +1,114 @@
+ * all keys after a batch has updated them.
+ *
+ * @tparam K Class of the state key
+ * @tparam S Class of the state data
+ * @tparam T Class of the emitted records
+ */
+@Experimental
+sealed abstract class EmittedRecordsDStream[K, S, T: ClassTag](
+    ssc: StreamingContext) extends DStream[T](ssc) {
+
+  /** Return a pair DStream where each RDD is the snapshot of the state of all the keys. */
+  def stateSnapshots(): DStream[(K, S)]
+}
+
+/** Internal implementation of the [[EmittedRecordsDStream]] */
+private[streaming] class EmittedRecordsDStreamImpl[
+    K: ClassTag, V: ClassTag, S: ClassTag, T: ClassTag](
+    trackStateDStream: TrackStateDStream[K, V, S, T])
+  extends EmittedRecordsDStream[K, S, T](trackStateDStream.context) {
+
+  override def slideDuration: Duration = trackStateDStream.slideDuration
+
+  override def dependencies: List[DStream[_]] = List(trackStateDStream)
+
+  override def compute(validTime: Time): Option[RDD[T]] = {
+    trackStateDStream.getOrCompute(validTime).map { _.flatMap[T] { _.emittedRecords } }
+  }
+
+  def stateSnapshots(): DStream[(K, S)] = {
+    trackStateDStream.flatMap[(K, S)] {
+      _.stateMap.getAll().map { case (k, s, _) => (k, s) }.toTraversable }
+  }
+}
+
+/**
+ * A DStream that allows per-key state to be maintains, and arbitrary records to be generated
+ * based on updates to the state.
+ *
+ * @param parent Parent (key, value) stream that is the source
+ * @param spec Specifications of the trackStateByKey operation
+ * @tparam K   Key type
+ * @tparam V   Value type
+ * @tparam S   Type of the state maintained
+ * @tparam T   Type of the eiitted records
+ */
+private[streaming] class TrackStateDStream[K: ClassTag, V: ClassTag, S: ClassTag, T: ClassTag](
+    parent: DStream[(K, V)], spec: StateSpecImpl[K, V, S, T])
+  extends DStream[TrackStateRDDRecord[K, S, T]](parent.context) {
+
+  persist(StorageLevel.MEMORY_ONLY)
+
+  private val partitioner = spec.getPartitioner().getOrElse(
+    new HashPartitioner(ssc.sc.defaultParallelism))
+
+  private val trackingFunction = spec.getFunction()
+
+  override def slideDuration: Duration = parent.slideDuration
+
+  override def dependencies: List[DStream[_]] = List(parent)
+
+  override val mustCheckpoint = true
+
+  /** Method that generates a RDD for the given time */
+  override def compute(validTime: Time): Option[RDD[TrackStateRDDRecord[K, S, T]]] = {
+    val prevStateRDD = getOrCompute(validTime - slideDuration).getOrElse {
+      TrackStateRDD.createFromPairRDD[K, V, S, T](
+        spec.getInitialStateRDD().getOrElse(new EmptyRDD[(K, S)](ssc.sparkContext)),
+        partitioner,
+        validTime.milliseconds
+      )
+    }
+    val newDataRDD = parent.getOrCompute(validTime).get
+    val partitionedDataRDD = newDataRDD.partitionBy(partitioner)
+    val timeoutThresholdTime = spec.getTimeoutInterval().map { interval =>
+      (validTime - interval).milliseconds
```

Comment:
Actually, in the current implementation, there is no way to make it exact. Since we dont maintain an timestamp-index in the data structure yet, we need to do full scans to find timedout records. And we do full scans infrequently, only at the timeof checkpointing. 

Nonetheless, i will document that the exact time of timing out isnt guaranteed.